pkg/a8rcloud/systema.go

```diff
@@ -0,0 +1,169 @@
+package a8rcloud
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/datawire/dlib/dgroup"
+	"github.com/datawire/dlib/dlog"
+)
+
+const (
+	ApiKeyHeader           = "X-Ambassador-Api-Key"
+	InstallIDHeader        = "X-Ambassador-Install-ID"
+	TrafficManagerIDHeader = "X-Telepresence-ManagerID"
+)
+
+const (
+	TrafficManagerConnName = "traffic-manager"
+	UserdConnName          = "userd"
+)
+
+type Closeable interface {
+	Close(ctx context.Context) error
+}
+
+type HeaderProvider interface {
+	GetAPIKey(ctx context.Context) (string, error)
+	GetInstallID(ctx context.Context) (string, error)
+	GetExtraHeaders(ctx context.Context) (map[string]string, error)
+}
+
+type ClientProvider[T Closeable] interface {
+	HeaderProvider
+	GetSystemaAddress(ctx context.Context) (string, error)
+	BuildClient(ctx context.Context, conn *grpc.ClientConn) (T, error)
+}
+
+type systemaPoolKey string
+
+func WithSystemAPool[T Closeable](ctx context.Context, poolName string, provider ClientProvider[T]) context.Context {
+	return context.WithValue(ctx, systemaPoolKey(poolName), &systemAPool[T]{Provider: provider, Name: poolName})
+}
+
+func GetSystemAPool[T Closeable](ctx context.Context, poolName string) SystemAPool[T] {
+	if p, ok := ctx.Value(systemaPoolKey(poolName)).(SystemAPool[T]); ok {
+		return p
+	}
+	return nil
+}
+
+type SystemAPool[T Closeable] interface {
+	Get(ctx context.Context) (T, error)
+	Done(ctx context.Context) error
```

Comment:
An alternative to a Get/Done pair could be:
```
   Do(context.Context, func(context.Context, T) error) error
```
that would use a private get/done pair internally.

Reply:
I'd love to do that, the only issue is the way UpdateIntercept currently expects the connection to be left dangling --it'd have to be changed to have a long-running goroutine that mirrors the intercept lifecycle

Reply:
Yeah, I understand that now. Never mind.